In [29]:
import torch 
import tensorflow as tf 
import numpy as np 
import math

In [30]:
data = [[12,3,4], [9,12,12]]
data = (3,4)
new_tensor=torch.rand(data)
new_tensor

tensor([[0.3195, 0.3463, 0.6963, 0.7241],
        [0.5216, 0.3449, 0.1661, 0.3334],
        [0.7025, 0.5203, 0.2448, 0.2514]])

In [31]:
# grad_fn 


a = torch.tensor(2.0, requires_grad=True)
b = torch.tensor(3.0, requires_grad=True)
x = torch.tensor(12.0, requires_grad=True)
y = a+b
z = x*y

print(f"grad_fn for z: {z.grad_fn}")
print(f"grad_fn for y: {y.grad_fn}")
print(f"grad_fn for x: {x.grad_fn}")

grad_fn for z: <MulBackward0 object at 0x0000028FEEACE110>
grad_fn for y: <AddBackward0 object at 0x0000028FEEACCAF0>
grad_fn for x: None


In [37]:
scores = torch.tensor([[10.,20.,30.], 
                       [5.,10.,15.]])

avg_per_assignment = scores.mean(dim=0) # 1 
avg_per_student = scores.mean(dim=1)    # 2

print(f"Avg per assignment: {avg_per_assignment}")
print(f"Avg per student: {avg_per_student}")


Avg per assignment: tensor([ 7.5000, 15.0000, 22.5000])
Avg per student: tensor([20., 10.])


In [38]:
scores.shape

torch.Size([2, 3])

| Student ↓ / Assignment → | A1 | A2 | A3 |
| ------------------------ | -- | -- | -- |
| Student 1                | 10 | 20 | 30 |
| Student 2                | 5  | 10 | 15 |


so, 

FOR 1. 

Calculation

    Assignment 1: (10 + 5) / 2 = 7.5
    
    Assignment 2: (20 + 10) / 2 = 15
    
    Assignment 3: (30 + 15) / 2 = 22.5


tensor([ 7.5, 15.0, 22.5 ])


FOR 2. 

Calculation

    Student 1: (10 + 20 + 30) / 3 = 20
    
    Student 2: (5 + 10 + 15) / 3 = 10

tensor([20., 10.])


In [52]:
## basic indexing 


x = torch.arange(12).reshape(3,4)

print(x)

col_2 = x[:,3] #get the third column (0-based indexing)
col_2

tensor([[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]])


tensor([ 3,  7, 11])

In [53]:
# dynamic selection 

scores = torch.tensor([
        [ -90,  0,  -20,  7],
        [ 8,  9, -109, 5]])

best_indices = torch.argmax(scores, dim=1)
best_indices

# the best index in the first row is 7, and in the second row here is 9. 

tensor([3, 1])

In [58]:
# expert selection 

data = torch.tensor([
        [ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11]
])

indices_to_select = torch.tensor([[2],[1],[3]])

selected_values = torch.gather(data, dim=1, index=indices_to_select)

selected_values

tensor([[ 2],
        [ 5],
        [11]])

In [60]:
N=10

D_in=1 
D_out=1

X = torch.randn(N, D_in)

true_W = torch.tensor([[2.0]])
true_b = torch.tensor(1.0)

y_true = X @ true_W + true_b + (torch.randn(N, D_out) * 0.1) # last - random noise. 

y_true

tensor([[-0.3530],
        [-0.0693],
        [ 0.9602],
        [-1.0544],
        [-1.1682],
        [ 1.4634],
        [-0.7360],
        [ 1.6622],
        [-2.1307],
        [ 2.5101]])

In [62]:
W = torch.randn(D_in, D_out, requires_grad=True)
b = torch.randn(1, requires_grad=True)

print(f"Initial W:{W}")
print(f"Initial b:{b}")

Initial W:tensor([[-0.5313]], requires_grad=True)
Initial b:tensor([0.5632], requires_grad=True)
